compiler/rustc_resolve/src/late.rs

```diff
@@ -1910,20 +1916,46 @@ impl<'a: 'ast, 'b, 'ast> LateResolutionVisitor<'a, 'b, 'ast> {
             add_bindings_for_ns(TypeNS);
         }
 
+        // Forbid shadowing lifetime bindings
+        for rib in self.lifetime_ribs.iter().rev() {
+            seen_lifetimes.extend(rib.bindings.iter().map(|(ident, _)| (*ident, false)));
+            if let LifetimeRibKind::Item = rib.kind {
+                break;
+            }
+        }
+
         for param in params {
             let ident = param.ident.normalize_to_macros_2_0();
             debug!("with_generic_param_rib: {}", param.id);
 
-            match seen_bindings.entry(ident) {
-                Entry::Occupied(entry) => {
-                    let span = *entry.get();
-                    let err = ResolutionError::NameAlreadyUsedInParameterList(ident.name, span);
-                    if !matches!(param.kind, GenericParamKind::Lifetime) {
-                        self.report_error(param.ident.span, err);
+            if let GenericParamKind::Lifetime = param.kind {
+                match seen_lifetimes.entry(ident) {
+                    Entry::Occupied(entry) => {
+                        let original = *entry.key();
+                        let orig_is_param = *entry.get();
+                        diagnostics::signal_lifetime_shadowing(
+                            self.r.session,
+                            original,
+                            param.ident,
+                            orig_is_param,
```

Comment:
`orig_is_param == true` is not exactly a shadowing, generic parameters do not create individual scopes, it's more like
```rust
mod m {
    struct S;
    struct S;
}
```

Can duplicate lifetime parameters
```rust
fn f<'a, 'a>() {}
```
be instead reported in the same way and using the same wording as other generic parameters like
```rust
fn f<T, T>() {}
fn f<const C: u8, const C: u8>() {}
```
?